javaslang/src/main/java/javaslang/collection/Iterator.java

```diff
@@ -2089,6 +2063,72 @@ public String toString() {
         }
     }
 
+    final class GroupedIterator<T> extends AbstractIterator<Seq<T>> {
+
+        private final Iterator<T> that;
+        private final int size;
+        private final int step;
+        private final int gap;
+        
+        private Seq<T> buffer = Vector();
+        private boolean filled = false;
+
+        GroupedIterator(Iterator<T> that, int size, int step) {
+            if (size < 1 || step < 1) {
+                throw new IllegalArgumentException("size (" + size + ") and step (" + step + ") must both be positive");
+            }
+            this.that = that;
+            this.size = size;
+            this.step = step;
+            this.gap = Math.max(step - size, 0);
+        }
+
+        @Override
+        public boolean hasNext() {
+            return filled || fill();
+        }
+
+        @Override
+        protected Seq<T> getNext() {
+            filled = false;
+            return buffer;
+        }
+
+        private boolean fill() {
+            if (!that.hasNext()) {
+                return false;
+            }
+
+            final int count = buffer.isEmpty() ? size : step;
+            final Seq<T> xs = Vector.ofAll(that.take(count));
+
+            if (xs.isEmpty()) {
+                return false;
+            }
+
+            final int prevSize = buffer.size();
+            final boolean isFirst = prevSize == 0;
+            final int len = xs.size();
+            final int howMany = Math.min(len, size);
+
+            if (howMany < 1 || !isFirst && len <= gap) {
+                return false;
+            }
+
+            final int available;
+            if (isFirst) {
+                available = len;
+            } else {
+                available = Math.min(howMany, len - gap);
+                buffer = buffer.drop(Math.min(step, prevSize));
+            }
+
+            buffer = buffer.appendAll(xs.takeRight(available));
+
+            return filled = true;
```

Comment:
this will always return true, right?
or did you mean `filled == true`? (in which case it needs some additional tests)